### PR TITLE
[MRG] Added TIDNet

### PR DESCRIPTION
**Pull request by HenrikBons**
As per issue #170
changes by @Ann-KathrinKiessner, @vytjan, @HenrikBons & @dcwil

### CODE REVIEW

File: braindecode/models/tidnet.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+from torch import nn
+from math import ceil
+from torch.nn import init
+from torch.nn.utils import weight_norm
+from .modules import Ensure4d, Expression
+
+import torch
+
+
+class Expand(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:10]: Is this unused? Than maybe remove it?

File: braindecode/models/tidnet.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    def __init__(self, axis=-1):
+        super().__init__()
+        self.axis = axis
+
+    def forward(self, x):
+        return x.unsqueeze(self.axis)
+
+
+class Flatten(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:19]: there is nn.Flatten from PyTorch, do you know why this is needed? try using nn.Flatten instead (with start_dim=1)

File: braindecode/models/tidnet.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    def forward(self, x):
+        return x.contiguous().view(x.size(0), -1)
+
+
+class BatchNormZG(nn.BatchNorm2d):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:24]: Maybe make this private with underscore, otherwise docstring would be necessary

File: braindecode/models/tidnet.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    def reset_parameters(self):
+        if self.track_running_stats:
+            self.running_mean.zero_()
+            self.running_var.fill_(1)
+        if self.affine:
+            self.weight.data.zero_()
+            self.bias.data.zero_()
+
+
+class ConvBlock2D(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:34]: Same here private with underscore

File: braindecode/models/tidnet.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    def forward(self, input, **kwargs):
+        res = input
+        input = self.conv(input, **kwargs)
+        input = self.dropout(input)
+        input = self.activation(input)
+        input = self.batch_norm(input)
+        return input + res if self.residual else input
+
+
+class DenseFilter(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:60]: Also private

File: braindecode/models/tidnet.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            nn.Conv2d(bottleneck * growth_rate, growth_rate, kernel,
+                      padding=tuple((k // 2 for k in kernel))),
+            nn.Dropout2d(drop_prob)
+        )
+
+    def forward(self, x):
+        return torch.cat((x, self.net(x)), dim=1)
+
+
+class DenseSpatialFilter(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:84]: Also private

File: braindecode/models/tidnet.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    def forward(self, x):
+        if len(x.shape) < 4:
+            x = x.unsqueeze(1).permute([0, 1, 3, 2])
+        x = self.net(x)
+        if self.collapse:
+            return self.channel_collapse(x).squeeze(-2)
+        return x
+
+
+class TemporalFilter(nn.Module):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:106]: Also private

File: braindecode/models/tidnet.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    Code adapted from: https://github.com/SPOClab-ca/ThinkerInvariance/blob/master/
+
+    References
+    ----------
+    .. [TIDNet] Kostas, D. & Rudzicz, F.
+        Thinker invariance: enabling deep neural networks for BCI across more people.
+        J. Neural Eng. 17, 056008 (2020).
+        doi: 10.1088/1741-2552/abb7a7.
+    """
+    def __init__(self, in_chans, n_classes, input_window_samples, s_growth=24, t_filters=32,
```
Comment (robintibor) [inline braindecode/models/tidnet.py:234]: you can probably remove the **kwargs?

File: braindecode/models/tidnet.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
+from torch import nn
+from math import ceil
+from torch.nn import init
+from torch.nn.utils import weight_norm
+from .modules import Ensure4d, Expression
+
+import torch
```
Comment (agramfort) [inline braindecode/models/tidnet.py:7]:
```suggestion
from math import ceil

import torch
from torch import nn
from torch.nn import init
from torch.nn.utils import weight_norm

from .modules import Ensure4d, Expression
```

import order should be from standard to specific ie std lib -> local relative imports

File: braindecode/models/tidnet.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    def forward(self, x):
+        if self.residual_style.lower() == 'netwise':
+            return self.net(x) + self.residual(x)
+        elif self.residual_style.lower() == 'dense':
+            for layer in self.net:
+                x = torch.cat((x, layer(x)), dim=1)
+            return x
+
+
+class _tidnet_features(nn.Module):
```
Comment (agramfort) [inline braindecode/models/tidnet.py:125]: class names should be CamelCase
```suggestion
class _TIDNetFeatures(nn.Module):
```

File: braindecode/models/tidnet.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        filter of length 75.
+    bottleneck : int
+        Bottleneck factor within Densefilter
+    summary : int
+        Output size of AdaptiveAvgPool1D layer. If set to -1, value will be calculated
+        automatically (input_window_samples // pooling).
+
+    Notes
+    -----
+    Code adapted from: https://github.com/SPOClab-ca/ThinkerInvariance/blob/master/
```
Comment (agramfort) [inline braindecode/models/tidnet.py:210]:
```suggestion
    Code adapted from: https://github.com/SPOClab-ca/ThinkerInvariance/
```

File: braindecode/models/tidnet.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+        self.classify = self._create_classifier(self.num_features, n_classes)
+
+    def _create_classifier(self, incoming, n_classes):
+        classifier = nn.Linear(incoming, n_classes)
+        init.xavier_normal_(classifier.weight)
+        classifier.bias.data.zero_()
+        return nn.Sequential(nn.Flatten(start_dim=1), classifier, nn.LogSoftmax(dim=-1))
+
+    def forward(self, x, **kwargs):
```
Comment (agramfort) [inline braindecode/models/tidnet.py:244]: docstring of public method missing

Comment (robintibor) [reply braindecode/models/tidnet.py:245]: do we really want docstring for forward methods? E.g., PyTorch doesn't do this: 
https://github.com/pytorch/pytorch/blob/bac4cfd54d44aa0fbc574e6561b878cb406762cc/torch/nn/modules/conv.py#L442

Neither does torchvision:
https://github.com/pytorch/vision/blob/d391a0e992a35d7fb01e11110e2ccf8e445ad8a0/torchvision/models/densenet.py#L215

Comment (robintibor) [reply braindecode/models/tidnet.py:245]: but I don't think we need **kwargs here? this could be removed?

Comment (agramfort) [reply braindecode/models/tidnet.py:245]: I would say here we could document the shapes of x for this network module? but I will not fight hard for it

Comment (robintibor) [reply braindecode/models/tidnet.py:245]: ok since you're not gonna fight for it and I think we should always have same accepted input shapes: #examples x #EEGchannels x #time (and optionally except empty last dim), we can just keep without docstring but remove **kwargs

File: braindecode/models/tidnet.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        self.activation = activation()
+        self.residual = residual
+
+        self.conv = nn.Conv2d(in_filters, out_filters, kernel, stride=stride, padding=padding,
+                              dilation=dilation, groups=groups, bias=not batch_norm)
+        self.dropout = nn.Dropout2d(p=drop_prob)
+        self.batch_norm = _BatchNormZG(out_filters) if residual else nn.BatchNorm2d(out_filters) if\
+            batch_norm else lambda x: x
+
+    def forward(self, input, **kwargs):
```
Comment (robintibor) [inline braindecode/models/tidnet.py:38]:
```suggestion
    def forward(self, input, **kwargs):
```
```suggestion
    def forward(self, input):
```

File: braindecode/models/tidnet.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        res = input
+        input = self.conv(input, **kwargs)
```
Comment (robintibor) [inline braindecode/models/tidnet.py:40]:
```suggestion
        input = self.conv(input,)
```

**Comment by robintibor** [comment]
two more **kwargs to go and we should be ready to merge!

**Comment by robintibor** [comment]
@kostasde Just for your information: We have also integrated TIDNet into braindecode, we made some adaptations to the code to fit our assumptions about input shapes etc., but hopefully it is all still functionally equivalent :) Thanks again for having your code online :)

**Comment by kostasde** [comment]
This is great, thanks for your efforts and for letting me know folks!
